File: Tests/CSharpCompiler/Roslyn/SideEffectAnalyzer.fs
// The MIT License (MIT)
// open SafetySharp.Modeling
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Roslyn

open System
open System.Linq
open System.Runtime.CompilerServices
open NUnit.Framework
open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax
open SafetySharp.Tests
open SafetySharp.Modeling.CompilerServices
open SafetySharp.CSharpCompiler.Roslyn
open SafetySharp.CSharpCompiler.Roslyn.Syntax

[<TestFixture>]
module SideEffectAnalyzer =

    let isSideEffectFree expression = 
        let compilation = TestCompilation (sprintf "
            struct S 
            {
                public static S operator+(S s) { return s; }
                public static S operator-(S s) { return s; }
                public static S operator!(S s) { return s; }
                public static S operator~(S s) { return s; }
                public static S operator+(S s1, S s2) { return s1; }
                public static S operator-(S s1, S s2) { return s1; }
                public static S operator*(S s1, S s2) { return s1; }
                public static S operator/(S s1, S s2) { return s1; }
                public static S operator%%(S s1, S s2) { return s1; }
                public static bool operator==(S s1, S s2) { return false; }
                public static bool operator!=(S s1, S s2) { return false; }
                public static bool operator>=(S s1, S s2) { return false; }
                public static bool operator>(S s1, S s2) { return false; }
                public static bool operator<=(S s1, S s2) { return false; }
                public static bool operator<(S s1, S s2) { return false; }
                public static S operator&(S s1, S s2) { return s2; }
                public static S operator|(S s1, S s2) { return s1; }
                public static bool operator^(S s1, S s2) { return false; }
                public static S operator<<(S s, int x) { return s; }
                public static S operator>>(S s, int x) { return s; }
                public static bool operator true(S s) { return false; }
                public static bool operator false(S s) { return false; }
            }

            class X 
            { 
                int _field;
                bool Property { get; set; }
                void M() 
                { 
                    S s = new S();
                    int a = 0, b = 0;
                    bool c = false, d = false;
                    decimal e = 1, f = 3;
                    var result = %s; 
                }
            }
        " expression)

        let methodDeclaration = compilation.FindMethodDeclaration "X" "M"
        let lastStatement = methodDeclaration.Body.Statements.Last () :?> LocalDeclarationStatementSyntax

        let analyzer = SideEffectAnalyzer compilation.SemanticModel
        analyzer.IsSideEffectFree lastStatement.Declaration.Variables.[0].Initializer.Value

    [<Test>]
    let ``throws when semantic model is null`` () =
        raisesArgumentNullException "semanticModel" (fun () -> SideEffectAnalyzer null |> ignore)

    [<Test>]
    let ``throws when expression is null`` () =
        let compilation = TestCompilation ""
        let analyzer = SideEffectAnalyzer compilation.SemanticModel
        raisesArgumentNullException "expression" (fun () -> analyzer.IsSideEffectFree null |> ignore)

    [<Test>]
    let ``literals are side effect free`` () =
        isSideEffectFree "1" =? true
        isSideEffectFree "false" =? true
        isSideEffectFree "1.3m" =? true

    [<Test>]
    let ``reads of local variables are side effect free`` () =
        isSideEffectFree "a" =? true
        isSideEffectFree "c" =? true
        isSideEffectFree "e" =? true

    [<Test>]
    let ``unary expressions on literals are side effect free`` () =
        isSideEffectFree "+1" =? true
        isSideEffectFree "-1.3m" =? true
        isSideEffectFree "~1" =? true
        isSideEffectFree "!true" =? true

    [<Test>]
    let ``non-increment/decrement unary expressions on local variables are side effect free`` () =
        isSideEffectFree "+b" =? true
        isSideEffectFree "-f" =? true
        isSideEffectFree "~a" =? true
        isSideEffectFree "!c" =? true

    [<Test>]
    let ``increment/decrement unary expressions on local variables are not side effect free`` () =
        isSideEffectFree "++b" =? false
        isSideEffectFree "++f" =? false
        isSideEffectFree "b++" =? false
        isSideEffectFree "f++" =? false
        isSideEffectFree "--b" =? false
        isSideEffectFree "--f" =? false
        isSideEffectFree "b--" =? false
        isSideEffectFree "f--" =? false

    [<Test>]
    let ``nested unary expressions check their operands`` () =
        isSideEffectFree "+(++b)" =? false
        isSideEffectFree "-(f--)" =? false

    [<Test>]
    let ``parenthesized expressions return whether their contained expression is side effect free`` () =
        isSideEffectFree "(1)" =? true
        isSideEffectFree "(++a)" =? false

    [<Test>]
    let ``non-increment/decrement unary expressions on local variables of user-defined type are not side effect free`` () =
        isSideEffectFree "+s" =? false
        isSideEffectFree "-s" =? false
        isSideEffectFree "~s" =? false
        isSideEffectFree "!s" =? false

    [<Test>]
    let ``binary expressions on literals are side effect free`` () =
        isSideEffectFree "1 + 1" =? true
        isSideEffectFree "2 - 1.3m" =? true
        isSideEffectFree "1 * 1" =? true
        isSideEffectFree "1 / 1" =? true
        isSideEffectFree "1 % 1" =? true
        isSideEffectFree "2 << 4" =? true
        isSideEffectFree "2 >> 1" =? true
        isSideEffectFree "true && false" =? true
        isSideEffectFree "true || false" =? true
        isSideEffectFree "true & false" =? true
        isSideEffectFree "true | false" =? true
        isSideEffectFree "true ^ false" =? true
        isSideEffectFree "1 == 1" =? true
        isSideEffectFree "1 != 1" =? true
        isSideEffectFree "1 >= 1" =? true
        isSideEffectFree "1 > 1" =? true
        isSideEffectFree "1 <= 1" =? true
        isSideEffectFree "1 < 1" =? true

    [<Test>]
    let ``binary expressions on local variables are side effect free`` () =
        isSideEffectFree "a + a" =? true
        isSideEffectFree "e - f" =? true
        isSideEffectFree "a * a" =? true
        isSideEffectFree "a / a" =? true
        isSideEffectFree "a % a" =? true
        isSideEffectFree "a << b" =? true
        isSideEffectFree "b >> a" =? true
        isSideEffectFree "c && d" =? true
        isSideEffectFree "c || d" =? true
        isSideEffectFree "c & d" =? true
        isSideEffectFree "c | d" =? true
        isSideEffectFree "c ^ d" =? true
        isSideEffectFree "a == a" =? true
        isSideEffectFree "a != a" =? true
        isSideEffectFree "a >= a" =? true
        isSideEffectFree "a > a" =? true
        isSideEffectFree "a <= a" =? true
        isSideEffectFree "a < a" =? true

    [<Test>]
    let ``nested binary expressions check their operands`` () =
        isSideEffectFree "1 + (++b)" =? false
        isSideEffectFree "a == ++a" =? false
        isSideEffectFree "c && (a++ == 1 || (++b == --a) == (c ^ (d == (--b > 1))))" =? false

    [<Test>]
    let ``binary expressions on local variables of user-defined type are not side effect free`` () =
        isSideEffectFree "s + s" =? false
        isSideEffectFree "s - s" =? false
        isSideEffectFree "s * s" =? false
        isSideEffectFree "s / s" =? false
        isSideEffectFree "s % s" =? false
        isSideEffectFree "s << b" =? false
        isSideEffectFree "s >> b" =? false
        isSideEffectFree "s && s" =? false
        isSideEffectFree "s || s" =? false
        isSideEffectFree "s & s" =? false
        isSideEffectFree "s | s" =? false
        isSideEffectFree "s ^ s" =? false
        isSideEffectFree "s == s" =? false
        isSideEffectFree "s != s" =? false
        isSideEffectFree "s >= s" =? false
        isSideEffectFree "s > s" =? false
        isSideEffectFree "s <= s" =? false
        isSideEffectFree "s < s" =? false

File: Source/Compiler/Roslyn/SideEffectAnalyzer.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.Compiler.Roslyn
{
	using System;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Symbols;
	using Utilities;

	/// <summary>
	///     Analyzes an expression to determine whether it has potential side effects or whether it is guaranteed to be
	///     side effect free. For instance:
	///     - 1 + 1: Side effect free
	///     - x + y: Potential side effect for variables of user-defined type with an overloaded + operator
	///     - f()  : Potential side effect
	///     - ++x  : Not side effect free
	/// </summary>
	public class SideEffectAnalyzer
	{
		/// <summary>
		///     The visitor instance that is used to analyze expressions for side effects.
		/// </summary>
		private readonly Visitor _visitor;

		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		/// <param name="semanticModel">The <see cref="SemanticModel" /> that should be used to check for overloaded operators.</param>
		public SideEffectAnalyzer(SemanticModel semanticModel)
		{
			Requires.NotNull(semanticModel, () => semanticModel);
			_visitor = new Visitor(semanticModel);
		}

		/// <summary>
		///     Determines whether the <paramref name="expression" /> is side effect free. If <c>false</c> is returned,
		///     <paramref name="expression" /> has at least one potential side effect.
		/// </summary>
		/// <param name="expression">The <see cref="ExpressionSyntax" /> that should be analyzed.</param>
		public bool IsSideEffectFree(ExpressionSyntax expression)
		{
			Requires.NotNull(expression, () => expression);
			return _visitor.Visit(expression);
		}

		/// <summary>
		///     Visits all kinds of expressions and determines whether they are side effect free. The visitor returns <c>true</c> to
		///     indicate that the analyzed expression is side effect free.
		/// </summary>
		private class Visitor : CSharpSyntaxVisitor<bool>
		{
			private readonly SemanticModel _semanticModel;

			public Visitor(SemanticModel semanticModel)
			{
				_semanticModel = semanticModel;
			}

			public override bool DefaultVisit(SyntaxNode node)
			{
				Assert.NotReached("Encountered an unexpected kind of syntax node: {0}.", node.Kind());
				return false;
			}

			public override bool VisitLiteralExpression(LiteralExpressionSyntax node)
			{
				return true;
			}

			public override bool VisitIdentifierName(IdentifierNameSyntax node)
			{
				var symbol = _semanticModel.GetSymbolInfo(node).Symbol;
				Assert.NotNull(symbol, "Expected a valid symbol.");

				if (symbol is ILocalSymbol || symbol is IFieldSymbol)
					return true;

				return false;
			}

			public override bool VisitPrefixUnaryExpression(PrefixUnaryExpressionSyntax node)
			{
				switch (node.Kind())
				{
					case SyntaxKind.UnaryPlusExpression:
					case SyntaxKind.UnaryMinusExpression:
					case SyntaxKind.BitwiseNotExpression:
					case SyntaxKind.LogicalNotExpression:
						var symbol = _semanticModel.GetSymbolInfo(node).Symbol as IMethodSymbol;
						Assert.NotNull(symbol, "Expected a valid method symbol.");

						if (symbol.IsBuiltInOperator(_semanticModel))
							return Visit(node.Operand);

						return false;
					case SyntaxKind.PreIncrementExpression:
					case SyntaxKind.PreDecrementExpression:
						return false;
					default:
						Assert.NotReached("Encountered an unexpected unary operator: {0}.", node.Kind());
						return false;
				}
			}

			public override bool VisitPostfixUnaryExpression(PostfixUnaryExpressionSyntax node)
			{
				switch (node.Kind())
				{
					case SyntaxKind.PostIncrementExpression:
					case SyntaxKind.PostDecrementExpression:
						return false;
					default:
						Assert.NotReached("Encountered an unexpected unary operator: {0}.", node.Kind());
						return false;
				}
			}

			public override bool VisitBinaryExpression(BinaryExpressionSyntax node)
			{
				switch (node.Kind())
				{
					case SyntaxKind.AddExpression:
					case SyntaxKind.SubtractExpression:
					case SyntaxKind.MultiplyExpression:
					case SyntaxKind.DivideExpression:
					case SyntaxKind.ModuloExpression:
					case SyntaxKind.LeftShiftExpression:
					case SyntaxKind.RightShiftExpression:
					case SyntaxKind.BitwiseOrExpression:
					case SyntaxKind.BitwiseAndExpression:
					case SyntaxKind.ExclusiveOrExpression:
					case SyntaxKind.EqualsExpression:
					case SyntaxKind.NotEqualsExpression:
					case SyntaxKind.LessThanExpression:
					case SyntaxKind.LessThanOrEqualExpression:
					case SyntaxKind.GreaterThanExpression:
					case SyntaxKind.GreaterThanOrEqualExpression:
						var symbol = _semanticModel.GetSymbolInfo(node).Symbol as IMethodSymbol;
						Assert.NotNull(symbol, "Expected a valid method symbol.");

						if (symbol.IsBuiltInOperator(_semanticModel))
							return Visit(node.Left) && Visit(node.Right);

						return false;
					case SyntaxKind.LogicalOrExpression:
					case SyntaxKind.LogicalAndExpression:
						// Roslyn doesn't provide a symbol for operators && and || for built-in types.
						// see also: https://roslyn.codeplex.com/workitem/327
						var leftType = _semanticModel.GetTypeInfo(node.Left).Type;
						var rightType = _semanticModel.GetTypeInfo(node.Left).Type;

						if (leftType.IsBuiltType(_semanticModel) && rightType.IsBuiltType(_semanticModel))
							return Visit(node.Left) && Visit(node.Right);

						goto case SyntaxKind.AddExpression;
					default:
						Assert.NotReached("Encountered an unexpected binary operator: {0}.", node.Kind());
						return false;
				}
			}

			public override bool VisitAssignmentExpression(AssignmentExpressionSyntax node)
			{
				return false;
			}

			public override bool VisitMemberAccessExpression(MemberAccessExpressionSyntax node)
			{
				return Visit(node.Name) && Visit(node.Expression);
			}

			public override bool VisitParenthesizedExpression(ParenthesizedExpressionSyntax node)
			{
				return Visit(node.Expression);
			}

			public override bool VisitInvocationExpression(InvocationExpressionSyntax node)
			{
				return false;
			}

			public override bool VisitThisExpression(ThisExpressionSyntax node)
			{
				return true;
			}

			public override bool VisitBaseExpression(BaseExpressionSyntax node)
			{
				return true;
			}

			public override bool VisitElementAccessExpression(ElementAccessExpressionSyntax node)
			{
				var symbol = _semanticModel.GetSymbolInfo(node).Symbol as IPropertySymbol;

				// If we get a symbol, this element access refers to a user-defined indexer. Always assume side effects.
				if (symbol != null)
					return false;

				// Otherwise, it is an array access, which is considered to be side effect free.
				return true;
			}

			public override bool VisitCastExpression(CastExpressionSyntax node)
			{
				var symbol = _semanticModel.GetSymbolInfo(node).Symbol as IMethodSymbol;

				// If we get a symbol, this cast is a user-defined conversion. Always assume side effects.
				if (symbol != null)
					return false;

				// Otherwise, it is a built-in conversion, which is considered to be side effect free.
				return true;
			}

			public override bool VisitObjectCreationExpression(ObjectCreationExpressionSyntax node)
			{
				return false;
			}
		}
	}
}